[IDE] Complete `nonisolated`, `unowned`, and access control attribute parameter

a7medev opened the pull request:
## Description

Added code completion support for:

- `nonisolated(unsafe)`
- `unowned(safe)`
- `unowned(unsafe)`
- `<access-control>(set)` (e.g. `private(set)`)

a7medev committed:
[IDE] Rename `CustomSyntaxAttributeKind` to `ParameterizedDeclAttributeKind`

a7medev committed:
[IDE] Complete `unowned`, `nonisolated`, `access-control` parameter

a7medev committed:
[IDE] Return Code Completion Status in Access Control Completion

a7medev committed:
[IDE] Recover from incomplete `unowned` and `nonisolated` in `isStartOfSwiftDecl`

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2882,11 +2912,43 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+        return makeParserCodeCompletionStatus();
+      }
+      
+      if (Tok2.is(tok::code_complete) && Tok.is(tok::identifier) &&
+                 !Tok.isContextualDeclKeyword()) {
+        consumeToken(tok::identifier);
+        if (CodeCompletionCallbacks) {
+          CodeCompletionCallbacks->completeDeclAttrParam(
+              ParameterizedDeclAttributeKind::AccessControl, 0, false);
+        }
+        consumeToken(tok::code_complete);
+        return makeParserCodeCompletionStatus();
+      }
```
Do you have a test case in mind that we need this code path for?

a7medev replied:
This is mainly to complete `set` in cases like this (the identifier after "(" is meaningless to the parser):
```
private(se|
```
But not in cases like (the identifier after "(" is another contextual keyword):
```
private( nonisolated|
```

This is mainly to allow code completion triggered while in the middle of writing `set` not just directly after "(".

ahoppen replied:
See https://github.com/swiftlang/swift/pull/79974#discussion_r1994287469, I don’t think this is needed.

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2882,11 +2912,43 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+      
+      const Token &Tok2 = peekToken();
+      
+      if (Tok.is(tok::code_complete)) {
+        if (CodeCompletionCallbacks) {
+          CodeCompletionCallbacks->completeDeclAttrParam(
+              ParameterizedDeclAttributeKind::AccessControl, 0, false);
+        }
+        consumeToken(tok::code_complete);
+        return makeParserCodeCompletionStatus();
+      }
```
I don't think this is ever hit. If `Tok` is a code completion token here, then we would have already returned from the `Tok.is(too::code_complete)` check above.

a7medev replied:
Nice catch, I guess I just accidentally duplicated that check 😅

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5616,33 +5691,61 @@ bool swift::isKeywordPossibleDeclStart(const LangOptions &options,
   P.consumeToken(tok::identifier);
   P.consumeToken(tok::l_paren);
+  
+  const Token &Tok2 = P.peekToken();
+  if (P.consumeIf(tok::code_complete)) {
+    // If a code complete token is present, recover from missing/incorrect argument and missing '('
+    P.consumeIf(tok::identifier);
+    P.consumeIf(tok::r_paren);
+    backtrack.cancelBacktrack();
+    return true;
+  }
+  
+  if (P.Tok.is(tok::identifier) && !P.Tok.isContextualDeclKeyword() && Tok2.is(tok::code_complete)) {
+    P.consumeToken(tok::identifier);
+    P.consumeToken(tok::code_complete);
+    
+    // If a code complete is present with a non-keyword token before it, recover from missing/incorrect argument and missing '('
+    if (!P.Tok.isContextualDeclKeyword()) {
+      P.consumeIf(tok::identifier);
+    }
+    P.consumeIf(tok::r_paren);
+    backtrack.cancelBacktrack();
+    return true;
+  }
```
Do you have examples which these checks and the switch to `CancellableBacktrackingScope` are needed for?

a7medev replied:
Consider the below example without the changes added:

```swift
nonisolated(|) var count = 0
```

If we trigger code completion, the parser will check if this is a start of a Swift declaration, it will encounter `nonisolated` and "(" so it will look the full `nonisolated(unsafe)` in `isParenthesizedNonisolated` which will return `false` so it will not consume `nonisolated(|)` and will end up consuming `nonisolated` alone at the end of `isStartOfSwiftDecl` then check if the rest (`(|) var count = 0`) is the start of a declaration which will return `false`. This code above ends up going through the code completion flow for a function argument and the code completion for `nonisolated(unsafe)` will not run.

---

**Why did we switch to `CancellableBacktrackScope`?**

The previous `isParenthesizedModifier` function used to check if this was a valid parenthesized modifier or not without consuming the tokens and the caller had to consume them if the condition succeeds which was easy as the tokens were always: `<identifier> <l_paren> <identifier> <r_paren>`.

The new logic can consume different tokens in different orders so if we're still using a normal `BacktrackScope` the caller would need to check again for the tokens and consume them (probably duplicating the code inside of this function).

To avoid this, I changed the function to consume the tokens if the result is `true` so the caller doesn't have to do consume the tokens again (hence the renaming from `isParenthesizedModifier` to `consumeIfParenthesizedModifier`) so I needed to cancel the backtracking and in such case and switched to `CancellableBacktrackScope`.

---

I'd love to know what you think, I'm open to ideas that can help us simplify this.

ahoppen reviewed (commented):
This is a great PR. Thank you, @a7medev 🙏🏽 

I left a few comments inline. The next steps would be to add some tests for these new completions in `test/IDE`.

a7medev commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2484,19 +2485,32 @@ static std::optional<Identifier> parseSingleAttrOptionImpl(
     return std::nullopt;
   }
   
+  if (P.Tok.is(tok::code_complete)) {
+    Status.setHasCodeCompletion();
+    codeCompletionCallback();
+  }
```
This is to allow code completion triggered while in the middle of writing `safe`/`unsafe` not just directly after "(" similar to what is done for `<access-modifier>(set)`.

ahoppen replied:
Did you check that this is actually needed. IIRC if you have `@unowned(saf|)`, we lex that as `@unowned(<completion token>)` and drop `saf`.

a7medev replied:
I just checked and found them separate `<identifier> <code-complete>`. Should we keep it as is or check this functionality in the lexer instead?

ahoppen replied:
Oh, sorry, I misremembered. You are correct but the convention for code completion is that if you have `@unowned(saf|)`, then the editor is expected to invoke code completion before `saf` (ie. as `@unowned(|saf)`) and use `saf` as the filter text for code completion items. For example, you can see this in SourceKit-LSP here: https://github.com/swiftlang/sourcekit-lsp/blob/9c867df103e37f11ac8e866a4aa7893a1a35377e/Sources/SourceKitLSP/Swift/CodeCompletion.swift#L23

a7medev replied:
Interesting, thanks for the reference. 🙏🏼 
I'll use this info to simplify the code.

a7medev replied:
I removed this check. 👍🏼

a7medev committed:
[IDE] Remove duplicate code & add parameter label

a7medev committed:
[IDE] Consume identifier and ')' after parenthesized modifier

This fixes an issue where valid parenthesized modifiers like nonisolated(unsafe) get misinterpreted as function calls.

a7medev committed:
[IDE] Remove unnecessary code complete checks

a7medev committed:
[IDE] Fix access control set completion as a declaration item

a7medev committed:
[Test] Add tests for `unowned`, `nonisolated`, and access control `set` completion

a7medev committed:
[IDE] Add parameter label comment for `HasLabel`

a7medev commented:
@ahoppen Thanks for the thorough review. 🙏🏼 I've addressed the added comments and added the tests.

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2473,8 +2473,9 @@ Parser::parseMacroRoleAttribute(
-    DeclAttrKind DK, bool allowOmitted, DiagRef nonIdentifierDiagnostic) {
+    Parser &P, SourceLoc Loc, SourceRange &AttrRange, StringRef AttrName, DeclAttrKind DK,
+    ParserStatus &Status, bool allowOmitted, DiagRef nonIdentifierDiagnostic,
+    llvm::function_ref<void()> codeCompletionCallback = {}) {
```
Sorry, I could have noticed this in the last review round but I think this could be simplified if, instead of the `codeCompletionCallback`, we would pass in a `std::optional<ParameterizedDeclAttributeKind>` and call the `CodeCompletionCallbacks` from in here. What do you think?

a7medev replied:
Yes, I think this simplifies the code a bit more. Fixed. 👍🏼

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5616,33 +5664,61 @@ bool swift::isKeywordPossibleDeclStart(const LangOptions &options,
+    return true;
+  }
+  
+  if (P.Tok.is(tok::identifier) && !P.Tok.isContextualDeclKeyword() && Tok2.is(tok::code_complete)) {
+    P.consumeToken(tok::identifier);
+    P.consumeToken(tok::code_complete);
+    
+    // If a code complete is present with a non-keyword token before it, recover from missing/incorrect argument and missing '('
+    if (!P.Tok.isContextualDeclKeyword()) {
+      P.consumeIf(tok::identifier);
+    }
+    P.consumeIf(tok::r_paren);
+    backtrack.cancelBacktrack();
+    return true;
+  }
```
Similar to https://github.com/swiftlang/swift/pull/79974#discussion_r1994321852, I think this isn’t actually needed. Or am I missing something?

a7medev replied:
I've removed it. 👍🏼

ahoppen reviewed (commented):
Thanks for adding the tests as well. Two more minor comments, otherwise this looks great to me.

a7medev committed:
[IDE] Pass `ParameterizedDeclAttributeKind` to `parseSingleAttrOption` rather than completion callback

a7medev committed:
[IDE] Remove invalid assertion in `completeDeclAttrParam`

a7medev commented:
@ahoppen I've addressed the added comments. Can you please re-review? 🙏🏼

ahoppen reviewed (approved):
Thank you. Looks good to me. I’ll trigger CI and if it passes, I’ll merge the PR.

ahoppen commented:
@swift-ci Please smoke test

a7medev commented:
@ahoppen I see the CI jobs for Linux and Windows are failing. I checked the output and found a similar error for both of them: `"error: cannot find 'PollIndexRequest' in scope"` in `BackgroundIndexingTests.swift` for SourceKit LSP. I'm not sure exactly what's the cause for it but looking at other PRs I found some CI failures with the same error so I'm suspecting it's either flaky or missing some kind of configuration. What do you think?

ahoppen commented:
The failure should be gone with https://github.com/swiftlang/swift-integration-tests/pull/156.

ahoppen commented:
@swift-ci Please smoke test Linux

ahoppen commented:
@swift-ci Please test Windows
